src/jdk.httpserver/share/classes/sun/net/httpserver/ServerImpl.java

```diff
@@ -223,44 +225,69 @@ public HttpsConfigurator getHttpsConfigurator () {
     }
 
     public final boolean isFinishing() {
-        return finished;
+        return finishedLatch.getCount() == 0;
     }
 
+    /**
+     * This method stops the server by adding a stop request event and
+     * waiting for the server until the event is triggered or until the maximum delay is triggered.
+     * <p>
+     * This ensures that the server is stopped immediately after all exchanges are complete. HttpConnections will be forcefully closed if active exchanges do not
+     * complete within the imparted delay.
+     *
+     * @param delay maximum delay to wait for exchanges completion, in seconds
+     */
```

Comment:
make synchronised to allow only one thread to execute stop at any one time

Reply:
Not sure that's the best solution. But if the `terminating` flag is already set then it means that another thread has initiated stop, and maybe in this case we should just jump to waiting on the latch - no need to post the event again or close the channel and wakeup the selector again. The part in `stop` that need synchronization already have it, so I don't think declaring stop as synchronized would be such a good idea. In particular it would prevent other threads to call getExchange() / endExchange() so I would be afraid it would prevent active exchanges from proceeding - unless we use wait/notify instead of a countdown latch - but that might be even more difficult to get right.